src/v/kafka/server/handlers/fetch.cc

```diff
@@ -671,6 +672,477 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+                if (!consensus) {
+                    results[i] = read_result(
+                      error_code::not_leader_for_partition);
+                    results[i].partition
+                      = ctx.requests[i].ktp().get_partition();
+                    co_return worker_result{
+                      .read_results = std::move(results),
+                      .errored_partitions = {i},
+                      .total_size = total_size,
+                    };
+                }
+                last_visible_indexes[i] = consensus->last_visible_index();
+            }
+            available_hwm_waiters->clear();
+
+            auto new_results = co_await fetch_ntps_in_parallel(
+              ctx.mgr,
+              ctx.srv.get_replica_selector(),
+              std::move(requests),
+              ctx.foreign_read,
+              ctx.deadline,
+              ctx.bytes_left,
+              ctx.srv.memory(),
+              ctx.srv.memory_fetch_sem());
+
+            // Override the older results of the partitions just queried with
+            // the new results.
+            for (size_t i = 0; i < results_map.size(); i++) {
+                auto res_i = results_map[i];
+                auto& r = results[res_i];
+                total_size -= r.data_size_bytes();
+                r = std::move(new_results[i]);
+                total_size += r.data_size_bytes();
+                if (r.error != error_code::none) {
+                    errored_partitions.push_back(res_i);
+                }
+            }
+
+            if (
+              total_size > ctx.min_bytes || !errored_partitions.empty()
+              || ctx.as.abort_requested()) {
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .errored_partitions = std::move(errored_partitions),
+                  .total_size = total_size,
+                };
+            }
+
+            // Register new call backs for partitions just read.
+            for (auto res_i : results_map) {
+                auto wait_opt = register_hwm_wait(
+                  ctx,
+                  ctx.requests[res_i].ktp(),
+                  model::next_offset(last_visible_indexes[res_i]),
+                  hwm_waiters_as);
+
+                if (!wait_opt) {
+                    results[res_i] = read_result(
+                      error_code::not_leader_for_partition);
+                    results[res_i].partition
+                      = ctx.requests[res_i].ktp().get_partition();
+                    co_return worker_result{
+                      .read_results = std::move(results),
+                      .errored_partitions = {res_i},
+                      .total_size = total_size,
+                    };
+                }
+
+                auto waiter = wait_opt.value().finally(
+                  [available_hwm_waiters, res_i] {
+                      available_hwm_waiters->push_back(res_i);
```

Comment:
question: why not just execute the logic needed to read here, then reset future, instead of pushing back an integer into a shared data structure for later querying? IMO kind of makes things confusing as we now have `available_hwm_waiters` and `hwm_waiters` which fundamentally represent the same.

Reply:
Good point, I'll switch `hwm_waiters` to be just a condition variable that any waiter can set when it's ready. Then `available_hwm_waiters` will be the only source of truth for which partitions have new data that can be read. That should simplify things a lot.